File: src/vibeusage/config/credentials.py
"""Credential file management for vibeusage."""

from __future__ import annotations

import os
import stat
from pathlib import Path

from vibeusage.config.paths import credentials_dir

# Provider CLI credential locations
PROVIDER_CREDENTIAL_PATHS: dict[str, str | list[str]] = {
    "claude": "~/.claude/.credentials.json",
    "codex": "~/.codex/auth.json",
    "gemini": "~/.gemini/oauth_creds.json",
    "copilot": "~/.config/github-copilot/hosts.json",
    "cursor": "~/.cursor/mcp-state.json",
}


def credential_path(provider_id: str, credential_type: str) -> Path:
    """Get the path for a provider's credential file."""
    return credentials_dir() / provider_id / f"{credential_type}.json"


def _expand_path(path: str) -> Path:
    """Expand user path and return Path object."""
    return Path(path).expanduser()


def find_provider_credential(provider_id: str) -> tuple[bool, str | None, Path | None]:
    """Check for provider credentials in vibeusage and provider CLIs.

    Returns:
        (found, source, path) - source is 'vibeusage', 'provider_cli', or None
    """
    from .settings import get_config

    config = get_config()
    config.get_provider_config(provider_id)

    # Check vibeusage storage first
    vibeusage_paths = [
        credential_path(provider_id, "oauth"),
        credential_path(provider_id, "session"),
        credential_path(provider_id, "apikey"),
    ]

    for cred_path in vibeusage_paths:
        if cred_path.exists():
            return True, "vibeusage", cred_path

    # Check provider CLI credentials if reuse is enabled
    if (
        config.credentials.reuse_provider_credentials
        and provider_id in PROVIDER_CREDENTIAL_PATHS
    ):
        provider_paths = PROVIDER_CREDENTIAL_PATHS[provider_id]
        if isinstance(provider_paths, str):
            provider_paths = [provider_paths]

        for provider_path in provider_paths:
            path = _expand_path(provider_path)
            if path.exists():
                return True, "provider_cli", path

    # Check environment variables for API keys
    env_var_names = {
        "claude": "ANTHROPIC_API_KEY",
        "codex": "OPENAI_API_KEY",
        "gemini": "GEMINI_API_KEY",
        "copilot": "GITHUB_TOKEN",
        "cursor": "CURSOR_API_KEY",
    }

    if env_var := env_var_names.get(provider_id):
        if os.environ.get(env_var):
            return True, "env", None

    return False, None, None


def write_credential(path: Path, content: bytes) -> None:
    """Securely write credential to file with 0o600 permissions."""
    path.parent.mkdir(parents=True, exist_ok=True)

    # Write to temp file first, then rename for atomicity
    temp_path = path.with_suffix(".tmp")
    temp_path.write_bytes(content)

    # Set restrictive permissions
    temp_path.chmod(stat.S_IRUSR | stat.S_IWUSR)  # 0o600

    # Atomic rename
    temp_path.replace(path)


def read_credential(path: Path) -> bytes | None:
    """Read credential file if it exists and has secure permissions."""
    if not path.exists():
        return None

    # Check permissions
    mode = path.stat().st_mode
    if mode & (stat.S_IRGRP | stat.S_IWGRP | stat.S_IROTH | stat.S_IWOTH):
        # File is readable/writable by group or others - unsafe
        return None

    return path.read_bytes()


def delete_credential(path: Path) -> bool:
    """Delete credential file.

    Returns:
        True if deleted, False if didn't exist
    """
    if not path.exists():
        return False

    path.unlink()
    return True


def check_credential_permissions(path: Path) -> bool:
    """Verify credential file has secure permissions (0o600 or stricter)."""
    if not path.exists():
        return True  # No file is secure

    mode = path.stat().st_mode
    # Check that only owner has read/write
    return not (mode & (stat.S_IRGRP | stat.S_IWGRP | stat.S_IROTH | stat.S_IWOTH))


def check_provider_credentials(provider_id: str) -> tuple[bool, str | None]:
    """Check if provider has valid credentials.

    Returns:
        (has_credentials, source) - source is 'vibeusage', 'provider_cli', 'env', or None
    """
    found, source, _ = find_provider_credential(provider_id)
    return found, source


def get_all_credential_status() -> dict[str, dict]:
    """Get credential status for all known providers."""
    status = {}
    for provider_id in PROVIDER_CREDENTIAL_PATHS:
        has_creds, source = check_provider_credentials(provider_id)
        status[provider_id] = {
            "has_credentials": has_creds,
            "source": source,
        }
    return status

File: vibeusage-go/internal/config/credentials.go
package config

import (
	"io/fs"
	"os"
	"path/filepath"
)

// ProviderCLIPaths maps provider IDs to their CLI credential file paths.
var ProviderCLIPaths = map[string][]string{
	"claude":  {"~/.claude/.credentials.json"},
	"codex":   {"~/.codex/auth.json"},
	"gemini":  {"~/.gemini/oauth_creds.json"},
	"copilot": {"~/.config/github-copilot/hosts.json"},
	"cursor":  {"~/.cursor/mcp-state.json"},
}

// ProviderEnvVars maps provider IDs to their environment variable names.
var ProviderEnvVars = map[string]string{
	"claude":  "ANTHROPIC_API_KEY",
	"codex":   "OPENAI_API_KEY",
	"gemini":  "GEMINI_API_KEY",
	"copilot": "GITHUB_TOKEN",
	"cursor":  "CURSOR_API_KEY",
}

func CredentialPath(providerID, credType string) string {
	return filepath.Join(CredentialsDir(), providerID, credType+".json")
}

func expandPath(p string) string {
	if len(p) > 1 && p[:2] == "~/" {
		home, err := os.UserHomeDir()
		if err != nil {
			return p
		}
		return filepath.Join(home, p[2:])
	}
	return p
}

// FindProviderCredential checks for credentials in vibeusage storage, provider CLIs, and env vars.
// Returns (found, source, path).
func FindProviderCredential(providerID string) (bool, string, string) {
	cfg := Get()
	_ = cfg // used for reuse_provider_credentials check

	// Check vibeusage storage first
	for _, credType := range []string{"oauth", "session", "apikey"} {
		p := CredentialPath(providerID, credType)
		if fileExists(p) {
			return true, "vibeusage", p
		}
	}

	// Check provider CLI credentials
	if cfg.Credentials.ReuseProviderCredentials {
		if paths, ok := ProviderCLIPaths[providerID]; ok {
			for _, raw := range paths {
				p := expandPath(raw)
				if fileExists(p) {
					return true, "provider_cli", p
				}
			}
		}
	}

	// Check environment variables
	if envVar, ok := ProviderEnvVars[providerID]; ok {
		if os.Getenv(envVar) != "" {
			return true, "env", ""
		}
	}

	return false, "", ""
}

func WriteCredential(path string, content []byte) error {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, content, 0o600); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

func ReadCredential(path string) ([]byte, error) {
	if !fileExists(path) {
		return nil, nil
	}
	info, err := os.Stat(path)
	if err != nil {
		return nil, err
	}
	mode := info.Mode()
	if mode&(fs.FileMode(0o077)) != 0 {
		// File is accessible by group/others — still read it but log warning
		// In practice we'll be permissive here since Go CLI tools often run differently
	}
	return os.ReadFile(path)
}

func DeleteCredential(path string) bool {
	if err := os.Remove(path); err != nil {
		return false
	}
	return true
}

func CheckProviderCredentials(providerID string) (bool, string) {
	found, source, _ := FindProviderCredential(providerID)
	return found, source
}

func GetAllCredentialStatus() map[string]map[string]any {
	status := make(map[string]map[string]any)
	for providerID := range ProviderCLIPaths {
		hasCreds, source := CheckProviderCredentials(providerID)
		status[providerID] = map[string]any{
			"has_credentials": hasCreds,
			"source":          source,
		}
	}
	return status
}

func IsFirstRun() bool {
	for providerID := range ProviderCLIPaths {
		hasCreds, _ := CheckProviderCredentials(providerID)
		if hasCreds {
			return false
		}
	}
	return true
}

func CountConfiguredProviders() int {
	count := 0
	for providerID := range ProviderCLIPaths {
		hasCreds, _ := CheckProviderCredentials(providerID)
		if hasCreds {
			count++
		}
	}
	return count
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
